sys/dev/sound/pci/hda/pin_patch.h

```diff
@@ -0,0 +1,100 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
+ *
+ * Copyright (c) 2018 Khamba Staring <k.staring@quickdecay.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ * $FreeBSD$
+ */
+#ifndef PIN_PATCH_H
+#define PIN_PATCH_H
+
+#include "hdac.h"
+
+#define PIN_SUBVENDOR(sv)		{ .id = sv }
+
+
+#define PIN_PATCH_STRING(n, patchstr) {	.nid = n,			\
+					.type = PIN_PATCH_TYPE_STRING,\
+					.patch.string = patchstr	\
+					}
+#define PIN_OVERRIDE(n, newvalue) {	.nid = n,			\
+					.type = PIN_PATCH_TYPE_OVERRIDE,\
+					.patch.override = newvalue	\
+					}
+#define PIN_PATCH_NOT_APPLICABLE(n)	PIN_OVERRIDE(n, 0x411111f0)
```

Comment:
What are these magic values? Could you please decode them?

Reply:
The magic numbers are numbers copied from Linux sources. Since these numbers 'worked for me', I assumed they'd work for others as well, seeing Linux is probably more widespread for audio-enabled setups. I wanted to port the Linux solutions with as little changes as possible to later translate the values, whenever possible, to readable strings. Step one, in short, would be to flat out port the Linux stuff as verbatim as possible-- it's enough work as it is seeing apparently confusing things in the Linux source as it is, step two would be to make them readable (it should be possible in a PIN_PATCH_TYPE_STRING manner). I've #define'd the most common magic numbers into something most probably correct, so I can easily translate them to PIN_PATCH_TYPE_STRING types later.
I'm no more happy with magic numbers than you, I think ;-)

Reply:
I'll get busy in the coming days working out *if* the Linux overrides can be completely represented in the string format which is already parsed by snd_hda(4) and if so, use PIN_PATCH_TYPE_STRING .

Reply:
ad-hoc pin config parser is at https://github.com/kstaring/hdaa_pin_config

Reply:
LSB problem was in my hdaa_pin_config tool, argh ;-)